### ISSUE TITLE
unnecessary_cast should ignore type aliases

### ISSUE THREAD

**Issue by bluenote10** [rust-lang/rust-clippy#8018]
Lint name: `unnecessary_cast`

I tried this code:

```rust
// somewhere in a central place (library)
pub type AudioFloat = f32;

// somewhere in client code, example 1:
let audio_buffer = vec![0.0 as AudioFloat; buffer_size];

// somewhere in client code, example 2:
fn double(x: AudioFloat) -> AudioFloat {
    // contrived example that uses a local constant
    let local_constant = 2.0 as AudioFloat;
    local_constant * x
}
```

Motivation of the code: Audio processing is typically done in `f32`. To allow for rare experiments with `f64`, the code is using a type alias to allow for easy switching (but full blown generic would be over-engineering).

**I expected to see this happen:** Clippy should not issue an `unnecessary_cast` when a type alias is involved. From a client code perspective this is not identical to a plain cast to `f32`, because if `AudioFloat` changes its type, the cast to `f32` is outdated and becomes an error.

**Instead, this happened:** Clippy complains:

```
casting float literal to `f32` is unnecessary
`#[warn(clippy::unnecessary_cast)]` on by default
...
...: try: `2.0_f32`
```

This fix suggestion defeats the whole purpose of using a type alias, we explicitly do not want to use `_f32` literals in the code.

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-unknown-linux-gnu
release: 1.56.1
LLVM version: 13.0.0
```

**Comment by giraffate** [comment]
It seems to be duplicate of https://github.com/rust-lang/rust-clippy/issues/8018.

**Comment by llogiq** [comment]
Agreed. Closing as duplicate.

**Comment by bluenote10** [comment]
> It seems to be duplicate of #8018.

#8018 is this issue itself :wink: Do you perhaps mean #6923?

**Comment by giraffate** [comment]
Oh, yes, I mean #6923.
